### PR TITLE
Fix: Bam and intermediate loom files have null file source (#3096)

### PR DESCRIPTION
**Pull request by danielsotirhos**
Author

- [x] PR title references issue
- [x] Title of main commit references issue
- [x] PR is linked to Zenhub issue

Author (reindex)

- [x] Added `r` tag to commit title                         <sub>or this PR does not require reindexing</sub>
- [x] Added `reindex` label to PR                           <sub>or this PR does not require reindexing</sub>

Author (freebies & chains)

- [x] Freebies are blocked on this PR                       <sub>or there are no freebies in this PR</sub>
- [x] Freebies are referenced in commit titles              <sub>or there are no freebies in this PR</sub>
- [x] This PR is blocked by previous PR in the chain        <sub>or this PR is not chained to another PR</sub>
- [x] Added `chain` label to the blocking PR                <sub>or this PR is not chained to another PR</sub>

Author (upgrading)

- [x] Documented upgrading of deployments in UPGRADING.rst  <sub>or this PR does not require upgrading</sub>
- [x] Added `u` tag to commit title                         <sub>or this PR does not require upgrading</sub>
- [x] Added `upgrade` label to PR                           <sub>or this PR does not require upgrading</sub>

Author (requirements, before every review)

- [x] Ran `make requirements_update`                        <sub>or this PR leaves requirements*.txt, common.mk and Makefile untouched</sub>
- [x] Added `R` tag to commit title                         <sub>or this PR leaves requirements*.txt untouched</sub>
- [x] Added `reqs` label to PR                              <sub>or this PR leaves requirements*.txt untouched</sub>

Author (before every review)

- [x] `make integration_test` passes in personal deployment <sub>or this PR does not touch functionality that could break the IT</sub>
- [x] Rebased branch on `develop`, squashed old fixups

Primary reviewer (after approval)

- [x] Commented in issue about demo expectations            <sub>or labelled issue as `no demo`</sub>
- [x] Decided if PR can be labeled `no sandbox`
- [x] PR title is appropriate as title of merge commit
- [x] Moved ticket to Approved column
- [x] Assigned PR to an operator

Operator (before pushing merge the commit)

- [x] Checked `reindex` label and `r` commit title tag
- [x] Checked that demo expectations are clear              <sub>or issue is labeled as `no demo`</sub>
- [x] Rebased and squashed branch
- [x] Sanity-checked history
- [x] Pushed PR branch to Github
- [x] Branch pushed to Gitlab                               <sub>or PR is labeled `no sandbox`</sub>
- [x] Build passes in sandbox and added `sandbox` label     <sub>or PR is labeled `no sandbox`</sub>
- [x] Started reindex in `sandbox`                          <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Checked for failures in `sandbox`                     <sub>or this PR does not require reindexing `sandbox`</sub>
- [x] Added PR reference to merge commit title
- [x] Collected commit title tags in merge commit title
- [x] Moved linked issue to Merged column
- [x] Pushed merge commit to Github

Operator (after pushing the merge commit)

- [x] Moved freebies to Merged column                       <sub>or there are no freebies in this PR</sub> 
- [ ] Shortened the PR chain                                <sub>or this PR is not the base of another PR</sub>
- [ ] Verified that `N reviews` labelling is accurate
- [ ] Pushed merge commit to Gitlab                         <sub>or this changes can be pushed later, together with another PR</sub>
- [ ] Deleted PR branch from Github and Gitlab

Operator (reindex) 

- [ ] Started reindex in `dev`                              <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Checked for failures in `dev`                         <sub>or this PR does not require reindexing or does not target `dev`</sub>
- [ ] Started reindex in `prod`                             <sub>or this PR does not require reindexing or does not target `prod`</sub>
- [ ] Checked for failures in `prod`                        <sub>or this PR does not require reindexing or does not target `prod`</sub>

Operator

- [ ] Unassigned PR

### CODE REVIEW

File: src/azul/plugins/metadata/hca/transform.py [hunk 1]
```diff
@@ -358,7 +358,16 @@ def for_id(cls, submitter_id: str) -> Optional['Submitter']:
...
-            return cls.for_id(file.submitter_id)
+            self = cls.for_id(file.submitter_id)
+            if self is None:
+                if (
+                    isinstance(file, api.AnalysisFile)
+                    and file.format in ('bam', 'loom')
+                    and any(isinstance(p, api.AnalysisProcess)
+                            for p in file.from_processes.values())
+                ):
+                    self = cls.by_title['DCP/2 Analysis']
```
Comment (hannes-ucsc) [inline src/azul/plugins/metadata/hca/transform.py:369]:
```suggestion
                    self = cls.dcp2
```

File: test/service/test_response.py [hunk 2]
```diff
@@ -1205,7 +1205,33 @@ def test_project_accessions_response(self):
...
+                                        'developmentStage': {
+                                            'human adult stage': {
+                                                'organ': {
+                                                    'brain': {
+                                                        'libraryConstructionApproach': {
+                                                            '10X v2 sequencing': [
+                                                                {
+                                                                    'uuid': '9ea1b54e-0679-4cdc-93ab-c584110168ac',
+                                                                    'version': '2019-02-14T192430.435350Z',
+                                                                    'name': 'possorted_genome_bam.bam',
```
Comment (hannes-ucsc) [inline test/service/test_response.py:1220]: A bam is not a matrix, AFAIK.

File: test/indexer/test_hca_indexer.py [hunk 3]
```diff
@@ -473,9 +473,9 @@ def test_contributor_matrices(self):
                 'matrices': [
                     {
                         'file': [
+                            # A supplementary file. The 'strata' value was provided in
```
Comment (hannes-ucsc) [inline test/indexer/test_hca_indexer.py:476]: This should be in a separate commit.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: src/azul/plugins/metadata/hca/transform.py [hunk 4]
```diff
@@ -986,6 +980,22 @@ def _samples(self, samples: Iterable[api.Biomaterial]) -> MutableJSON:
                 assert False, sample
         return result
 
+    def _is_dcp_matrix(self, file: api.File) -> bool:
+        submitter = Submitter.for_file(file)
```
Comment (amarjandu) [inline src/azul/plugins/metadata/hca/transform.py:984]: can `submitter.category_for_file()` be used here? 
Thinking about ways to de-duplicate.

**Review by amarjandu — CHANGES_REQUESTED** [review]
Looks pretty good, small question....

**Review by amarjandu — APPROVED** [review]

File: src/azul/plugins/metadata/hca/transform.py [hunk 5]
```diff
@@ -986,6 +989,14 @@ def _samples(self, samples: Iterable[api.Biomaterial]) -> MutableJSON:
                 assert False, sample
         return result
 
+    def _is_dcp_matrix(self, file: api.File) -> bool:
+        category = Submitter.category_for_file(file)
+        return category == SubmitterCategory.internal
```
Comment (hannes-ucsc) [inline src/azul/plugins/metadata/hca/transform.py:994]: I think this method name is a bit misleading. A file with from an internal submitter aka source isn't automatically a matrix file. It used to be but we'll end up being wholly confused in the future unless we establish and equality between the Submitter enum and the file_source schema enum.

Also, we shouldn't hide our enum behind wrapper functions. The general approach should be to determine an enum instance once and them interrogate it.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
I sketched something out. See REVIEW commit.

File: src/azul/plugins/metadata/hca/transform.py [hunk 6]
```diff
@@ -737,12 +757,15 @@ def _file_types(cls) -> FieldTypes:
...
+
     def _file(self, file: api.File, related_files: Iterable[api.File] = ()) -> MutableJSON:
-        file_source = Submitter.title_for_file(file)
-        if file_source:
-            is_intermediate = False
-        elif any('matrix' in c.lower() for c in file.content_description):
-            is_intermediate = True
+        submitter = Submitter.for_file(file)
+        if self._is_matrix(file):
+            is_intermediate = file.submitter_id is None
```
Comment (hannes-ucsc) [inline src/azul/plugins/metadata/hca/transform.py:768]:
```suggestion
            # As per DCP/2 System Design, intermediate matrices generated by DCP/2 analysis do not carry a submitter ID. Also see Submitter.for_file
            is_intermediate = file.submitter_id is None
```

File: test/indexer/data/f0731ab4-6b80-4eed-97c9-4984de81a47c.metadata.json [hunk 7]
```diff
@@ -190,7 +190,14 @@
         "describedBy": "https://schema.humancellatlas.org/type/file/6.0.0/analysis_file",
         "file_core": {
             "file_name": "merged-cell-metrics.csv.gz",
-            "format": "unknown"
+            "format": "unknown",
+            "content_description": [
+                {
+                    "ontology": "data:3917",
+                    "ontology_label": "count matrix",
```
Comment (hannes-ucsc) [inline test/indexer/data/f0731ab4-6b80-4eed-97c9-4984de81a47c.metadata.json:197]: Isn't the label `Count matrix`?

File: test/indexer/data/f0731ab4-6b80-4eed-97c9-4984de81a47c.metadata.json [hunk 8]
```diff
@@ -190,7 +190,14 @@
         "describedBy": "https://schema.humancellatlas.org/type/file/6.0.0/analysis_file",
         "file_core": {
             "file_name": "merged-cell-metrics.csv.gz",
-            "format": "unknown"
+            "format": "unknown",
+            "content_description": [
```
Comment (hannes-ucsc) [inline test/indexer/data/f0731ab4-6b80-4eed-97c9-4984de81a47c.metadata.json:194]: Is this realistic? Did DCP/2 Analysis provide this field.

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]

File: test/indexer/data/f0731ab4-6b80-4eed-97c9-4984de81a47c.metadata.json (hunk 8 above)
Comment (danielsotirhos) [reply test/indexer/data/f0731ab4-6b80-4eed-97c9-4984de81a47c.metadata.json:194]: This `content_description` value was copied from another canned bundle. I checked a couple snapshots and found a more realistic value `Gene expression matrix` (data:3112) to associate with the `csv.gz` analysis file.

I wasn't able to find any snapshots containing a `.npz` analysis file with `content_description` set, so I also removed the `content_description` I had added to the file `sparse_counts.npz` and instead added it to the other `csv.gz` analysis file in the bundle so the test cases would have 2 analysis file & 1 supplementary file matrices to test with.

Comment (hannes-ucsc) [reply test/indexer/data/f0731ab4-6b80-4eed-97c9-4984de81a47c.metadata.json:194]: But to confirm. Do the analysis subgraphs set the content_description on the actual loom files? Or is this a organic CGM subgraph?

**Review by hannes-ucsc — CHANGES_REQUESTED** [review]
Just this on [question](https://github.com/DataBiosphere/azul/pull/3134#discussion_r659105302) for PL.

The rest looks good. We should be good to move _is_matrix to the API at this point. No bump.

**Review by hannes-ucsc — APPROVED** [review]
